language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -121,29 +122,49 @@ data CekUserError
     | CekEvaluationFailure -- ^ Error has been called.
     deriving (Show, Eq)
 
+{- Note [Being generic over @term@ in 'CekM']
+We have a @term@-generic version of 'CekM' called 'CekCarryingM', which itself requires a
+@term@-generic version of 'CekEvaluationException' called 'CekEvaluationExceptionCarrying'.
+
+The point is that in many different cases we can annotate an evaluation failure with a 'Term' that
+caused it. Originally we were using 'CekValue' instead of 'Term', however that meant that we had to
+ignore some important cases where we just can't produce a 'CekValue', for example if we encounter
+a free variable, we can't turn it into a 'CekValue' and report the result as the cause of the
+failure, which is bad. 'Term' is strictly more general than 'CekValue' and so we can always
+
+1. report things like free variables
+2. report a 'CekValue' turned into a 'Term' via 'dischargeCekValue'
+
+We need the latter, because the constant application machinery, in the context of the CEK machine,
+expects a list of 'CekValue's and so in the event of failure it has to report one of those
+arguments, so we have no option but to call the constant application machinery with 'CekValue'
+being the cause of a potential failure. But as mentioned, turning a 'CekValue' into a 'Term' is
+no problem and we need that elsewhere anyway, so we don't need any extra machinery for calling the
+constant application machinery over a list of 'CekValue's and turning the cause of a possible
+failure into a 'Term', apart from the straightforward generalization of 'CekM'.
+-}
+
+-- | The CEK machine-specific 'EvaluationException', parameterized over @term@.
+type CekEvaluationExceptionCarrying term =
+    EvaluationException UnknownDynamicBuiltinNameError CekUserError term
+
+-- See Note [Being generic over @term@ in 'CekM'].
+-- | A generalized version of 'CekM' carrying a @term@.
+-- 'State' is inside the 'ExceptT', so we can get it back in case of error.
+type CekCarryingM term uni =
+    ReaderT (CekEnv uni) (ExceptT (CekEvaluationExceptionCarrying term) (State ExBudgetState))
+
 -- | The CEK machine-specific 'EvaluationException'.
-type CekEvaluationException uni =
-    EvaluationException UnknownDynamicBuiltinNameError CekUserError (CekValue uni)
+type CekEvaluationException uni = CekEvaluationExceptionCarrying (WithMemory Term uni)
+
+-- | The monad the CEK machine runs in.
+type CekM uni = CekCarryingM (WithMemory Term uni) uni
```

Comment:
Do people think we need to report memory-annotated terms? I think it would be a little bit nicer if the machines were accepting unannotated terms and returning unannotated terms, regardless of whether the term is the cause of a failure or a result.

Reply:
I agree with this. If anything, you'd like to to return the original annotation like the location! But no point returning the memory annotation.